### PR TITLE
Render dynamic-workflow run tree on the nested DOM (#174 PR3)

### LINKED ISSUE #174: Add support for dynamic workflows
See [Orchestrate subagents at scale with dynamic workflows - Claude Code Docs](https://code.claude.com/docs/en/workflows).

We should have most of the pieces in place (sub-agents, teammates, etc.), so I suppose it'll be about the orchestrator script... We'll see once we have a transcript example to analyze.

### PR DESCRIPTION
**Pull request by cboos**
Final PR for #174 (dynamic-workflow support). Builds on PR1 (parse) and PR2
(Workflow tool-input rendering): splices each parsed `WorkflowRun` — phases →
agents → each agent's side-channel transcript — into the message tree at its
`Workflow` tool_use site, rendered on the existing nested DOM.

## What this adds

- **Run-tree splice.** `_splice_workflow_runs` (the last pass in
  `generate_template_messages`) attaches a self-contained sub-tree under each
  Workflow tool_use: a `WorkflowPhaseMessage` per phase, a `WorkflowAgentMessage`
  per agent (model / state / tokens / tool-calls + result — StructuredOutput
  dicts pretty-printed as JSON, plain strings as Markdown), and each agent's
  side-channel transcript grafted beneath it. Renders in both HTML and Markdown.
- **Synthetic-node indices** are allocated via `ctx.register` (`len(messages)`),
  an inherently session-wide monotonic allocator, so multiple/concurrent
  workflows in one session never collide. Grafted side-channel nodes are
  re-registered into the main context and their pairing references remapped.
- **Timeline parity.** Dedicated `workflow_phase` / `workflow_agent` lanes plus
  detection branches placed before the generic tool_use match (the cards carry
  `tool_use` for transcript-filter visibility).
- **Single-file rendering.** `claude-code-log <session>.jsonl` now discovers the
  session's sibling `subagents/workflows/` runs and splices them, matching the
  directory-load behavior.
- **Pagination-robust linkage.** Run↔tool_use linkage is resolved once at
  full-session scope (before pagination), so a Workflow tool_use links to its
  run even when its tool_result lands on a different page.

## Verification

- Validated end-to-end on a real ~76k-message project: the run renders its
  phases + 42 agent cards via the directory, single-file, and
  cache+pagination paths.
- New tests cover the splice structure, side-channel grafting, index
  uniqueness, HTML + Markdown rendering, single-file rendering, the
  pagination-boundary case, and a Playwright fold test for the phase nodes.
- Non-workflow output is byte-identical (snapshot diff is only the added
  workflow CSS rules + timeline detection JS); full unit suite, browser tests,
  and pyright all green.

Closes #174.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Dynamic workflow phases and agents are rendered as interactive cards (metadata, model/tokens, results) and are spliced into transcripts.
  * Workflow header resolution prefers snapshot data; single-file transcripts now discover and render linked workflow runs.

* **Bug Fixes**
  * Improved cross-page/linkage handling so workflow run links are resolved at full-session scope to avoid pagination breakage.

* **Tests**
  * Added comprehensive rendering tests including browser folding behavior.

* **Documentation**
  * Updated design docs describing the splice/render strategy and verification.

### CODE REVIEW

File: claude_code_log/html/tool_formatters.py [hunk 1]
```diff
@@ -1223,6 +1227,74 @@ def format_workflow_input(workflow_input: WorkflowToolInput) -> str:
...
+        parts.append(f"<div class='workflow-agent-meta'>{''.join(meta_bits)}</div>")
+
+    result = content.result
+    if isinstance(result, (dict, list)):
+        # Pretty-print + JSON-highlight via the shared async-result helper
+        # (it re-parses the `{"`/`[`-shaped text and indents it).
+        parts.append(
+            render_async_result_body(
+                json.dumps(result, ensure_ascii=False), "workflow-agent-result"
+            )
```
[the remark this answers is not in the document]
Comment (cboos) [reply claude_code_log/html/tool_formatters.py:1295]: (Claude) Good catch — real divergence. `format_workflow_agent_content` routed both dict and list through `render_async_result_body`, whose JSON heuristic only fires on `{"`, so a list-shaped (`[...]`) StructuredOutput result fell through to the markdown path — while the Markdown renderer already fenced both dict and list as JSON. Fixed in adf9cee: dict/list now JSON-highlight directly via `render_file_content_collapsible` (a real dict/list always serializes to valid JSON). Added a test asserting Pygments highlight in HTML and a ```json fence in Markdown for a list result.

File: work/dynamic-workflow-pr3-design.md [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+   fixtures (workflow_basic isn't a snapshot fixture).
+
+**Tests**: extend `test_workflow_rendering.py` — directory render of
+workflow_basic shows phase/agent cards nested under the Workflow tool_use; each
+agent's side-channel entries present beneath it; HTML + Markdown; a fold
+Playwright test. Snapshot regen serial (`-n0`); `just ci` green.
+
+**EXACT NEXT ACTION**: implement A→B→C (node types + registry + formatters +
+CSS) first and unit-test rendering a synthesized node; THEN D→E (the splice +
+re-index) — the highest-risk part; THEN F (timeline) + tests.
```
[the remark this answers is not in the document]
Comment (cboos) [reply work/dynamic-workflow-pr3-design.md]: (Claude) Fixed in adf9cee — replaced the stale "EXACT NEXT ACTION" forward-plan with a DONE/as-built status note (A→F all implemented + the two post-impl additions; just ci green; PR open). The map is retained as as-built reference.

File: work/dynamic-workflow-pr3-design.md [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+`_dispatch_title` (L4471). Add to BOTH `HtmlRenderer` and `MarkdownRenderer`:
+- `format_WorkflowPhaseMessage` / `title_WorkflowPhaseMessage` (header card:
+  title + detail + "N agents").
+- `format_WorkflowAgentMessage` / `title_WorkflowAgentMessage` (label/model/
+  state/tokens; body = result — dict → JSON-pygmentize (reuse
+  `render_async_result_body`), str → markdown).
+
+**D. Splice pass** (`renderer.py`, new `_splice_workflow_runs(root_messages,
+ctx)` called AFTER `_build_message_tree` (~L821), BEFORE
+`_link_async_notifications`; guard: only when a Workflow tool_use has
```
[the remark this answers is not in the document]
Comment (cboos) [reply work/dynamic-workflow-pr3-design.md]: (Claude) Fixed in adf9cee — the D-header now states the splice runs LAST in `generate_template_messages` (after `_link_task_id_consumers`), matching both the implementation and the §1/step-D.1 rationale (the `ctx.register` appends must not perturb earlier ctx.messages-iterating passes). Removed the contradictory "BEFORE _link_async_notifications" wording.
